### PR TITLE
ci: added qemu ARM executable testing

### PR DESCRIPTION
**Pull request by edtubbs**
Addresses https://github.com/dogecoin/dogecoin/issues/2888 with qemu for benchmark and test_dogecoin on arm as well as native macos.

### CODE REVIEW

**Comment by patricklodder** [comment]
Thanks! ❤️

I have 3 considerations:

1. Timing. We only have a couple of days of `1.14.6-dev` left so this will not see a lot of runs before release - I was thinking that we may want to base this on `1.14.7` instead so that we can stabilize it while the cycle is still young? No action needed - if we'd opt to do this, we'll just change the base branch of the PR once we have `1.14.7-dev`.
2. Concept ACK on the ARM/qemu part. This looks great!
3. A soft concept NAK on the macOS native part because we currently test no other builds without using the depends system, so this introduces maintenance pressure, especially since Homebrew is a big factor of uncertainty and imho shouldn't become a blocker on the integration path for every PR as long as we have debt towards dependencies (QT, OpenSSL) For 1.21 it works right now, but if we accumulate too much debt on there I expect that it will become a source of failures. Also note that I had a working build process on both my Intel and M1 macs when we merged #2745 but within 3 weeks after that, before I had finished the doc in #2686, updates to brew packages interfered with the process I wrote out.

Perhaps we can address the ARM/qemu and the macOS separately by splitting up this PR into 2? Would make it easier to merge the ARM part.

**Comment by edtubbs** [comment]
Ok, that sounds good. I've removed macos native, and I'll squash those commits.  I was able to build depends on macos native, but had to update boost.  Unfortunately that produces a glibc compatibility issue in other builds. 
https://github.com/edtubbs/dogecoin/runs/7271877876

**Comment by patricklodder** [comment]
Thanks!

> I was able to build depends on macos native, but had to update boost. Unfortunately that produces a glibc compatibility issue in other builds.

Perhaps we're able to patch the current boost 1.63 instead for macOS, to prevent needing a higher minimum glibc? We can investigate this separately.

PS: for 1.21, the minimum glibc compatibility requirements are per architecture: https://github.com/dogecoin/dogecoin/blob/a8313b847fcc0b01477aa8941f5e060b12db59c1/contrib/devtools/symbol-check.py#L78-L84

and we depend less on boost, more on stdlib.

**Comment by edtubbs** [comment]
Sure thing!
> Perhaps we're able to patch the current boost 1.63 instead for macOS, to prevent needing a higher minimum glibc? We can investigate this separately.
> 
> PS: for 1.21, the minimum glibc compatibility requirements are per architecture:
> 
> https://github.com/dogecoin/dogecoin/blob/a8313b847fcc0b01477aa8941f5e060b12db59c1/contrib/devtools/symbol-check.py#L78-L84
> 
> and we depend less on boost, more on stdlib.

Agreed, we should be able to patch boost 1.63 for macOS. 

I saw that in 1.21, too.  Awesome!

**Comment by patricklodder** [comment]
There is a patch in #2399 but it's very big for just a CI target. Bitcoin Core eventually did another fix where they split `b2` from `boost`, which also helps with cross-compilation, that may be beneficial to look into.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -245,15 +247,77 @@ jobs:
...
+              ;;
+              "i686-w64-mingw32")
+              make check $MAKEJOBS VERBOSE=1
+              qa/pull-tester/install-deps.sh
+              qa/pull-tester/rpc-tests.py --coverage
+              ;;
+              "x86_64-apple-darwin11")
+              make check $MAKEJOBS VERBOSE=1
+              ./src/test/test_dogecoin
+              ;;
```
Comment (patricklodder) [inline .github/workflows/ci.yml:309]: probably want to keep the macOS one out for the time being

**Review by patricklodder — CHANGES_REQUESTED** [review]
I have recorded my suggestions in a review branch: [patricklodder/review-suggestions-3025](https://github.com/patricklodder/dogecoin/commits/review-suggestions-3025)

### Running tests within the bench job

There's a C&P error on the bench job where unit tests are ran. Needs to be removed. 

See: https://github.com/patricklodder/dogecoin/commit/11b36fe37ef2fcdd59f24bf6c3f7f7a3f0944689

### Usefulness of emulated benchmarks

I ran this a couple of times and it looks like when emulated, the intrinsics are making the benchmark run slower, not faster, making the results misleading. On natively benched x86_64, the results however do seem to be as expected. I've ran it through a couple of times, but see below for the results from the latest CI run on this PR:

#### Emulated aarch64 benchmarks

| benchmark  |   aarch64 | aarch64-v8 | aarch64-v82 |
| :--------- | --------: | ---------: | ----------: |
| SHA1       | 0.0057948 |  0.0074411 |   0.0149350 |
| SHA256     | 0.0116468 |  0.0127086 |   0.0243240 |
| SHA256-32b | 1.1271959 |  1.1414134 |   1.9908715 |
| SHA512     | 0.0062912 |  0.0052691 |   0.0331773 |
| Scrypt     | 0.0010026 |  0.0008485 |   0.0011126 |

#### Native x86_64 benchmarks

| benchmark  |    x86_64 | x86_64 AVX2 |
| :--------- | --------: | ----------: |
| SHA1       | 0.0021777 |   0.0017780 |
| SHA256     | 0.0054168 |   0.0034975 |
| SHA256-32b | 0.3740650 |   0.2524940 |
| SHA512     | 0.0035571 |   0.0022788 |
| Scrypt     | 0.0003361 |   0.0003027 |

Therefore, I think it is better for now to leave out the emulated benchmark.

See: https://github.com/patricklodder/dogecoin/commit/b20f44eaaf4096e5c797ca6abbd06bbe6c977270

### Test implementation matrix

Currently we add a matrix under the matrix. It is probably less complicated and easier to maintain/debug to just use the top-level matrix to specify what we test. 

This would have the added benefit that we don't have to rely on graceful disabling of the qa tests for `linux-nowallet` and windows builds (and save some time building the scrypt python dependency on these runs)

See: https://github.com/patricklodder/dogecoin/commit/1fdd39c87e4564a433296918554b8aa6d2bb348a

### Conclusion

Concept ACK. I think the above is in order of importance to me, and only the C&P error is blocking, the rest is cleanup and we can always fix that later.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (edtubbs) [reply .github/workflows/ci.yml:309]: Awesome!  I concur with your recommendations.  I can merge your commits into my branch if that's alright?  We'll continue to work on automated testing for macos.

Comment (patricklodder) [reply .github/workflows/ci.yml:309]: > I can merge your commits into my branch if that's alright?

Of course!

**Comment by patricklodder** [comment]
As hinted above, I'm tagging this 1.14.7 and marking it as blocked (because we don't have a 1.14.7-dev branch yet), to keep focus on 1.14.6 release blockers for the next couple of days. I think we can pick this up next week ❤️

**Comment by patricklodder** [comment]
CI failures: GH Actions seems to have some issues with Azure's apt mirror - I'll restart the failing jobs once the running ones are done.

**Review by patricklodder — APPROVED** [review]
utACK. Need a second because I contributed to this.

**Comment by patricklodder** [comment]
@chromatic could you have a look to make sure I didn't mess anything up in my suggestions to Ed? Thx ❤️

**Review by chromatic — APPROVED** [review]
ACK from me, though CI is the real test. I've read through the changes and the rationales, and they all make sense to me. Let's get this testing!

**Pull request merged**